tests/sanity/sanity_test.go

```diff
@@ -0,0 +1,396 @@
+	region           = "us-west-2"
+	availabilityZone = "us-west-2a"
+)
+
+var (
+	disks            = make(map[string]*cloud.Disk)
+	snapshots        = make(map[string]*cloud.Snapshot)
+	snapshotNameToID = make(map[string]string)
+	mounts           = make(map[string]string)
+	fakeMetaData     = &metadata.Metadata{
+		InstanceID: instanceID,
+		Region:     region,
+	}
+	outpostArn = &arn.ARN{
+		Partition: "aws",
+		Service:   "outposts",
+		Region:    "us-west-2",
+		AccountID: "123456789012",
+		Resource:  "op-1234567890abcdef0",
+	}
+)
+
+func TestSanity(t *testing.T) {
+	defer func() {
+		if r := recover(); r != nil {
+			t.Errorf("Test panicked: %v", r)
+		}
+	}()
+
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	tmpDir, err := os.MkdirTemp("", "csi-sanity-")
+	if err != nil {
+		t.Fatalf("Failed to create sanity temp working dir: %v", err)
+	}
+	defer os.RemoveAll(tmpDir)
+
+	defer func() {
+		if err = os.RemoveAll(tmpDir); err != nil {
+			t.Fatalf("Failed to clean up sanity temp working dir %s: %v", tmpDir, err.Error())
+		}
+	}()
+
+	endpoint := fmt.Sprintf("unix:%s/csi.sock", tmpDir)
+	mountPath := path.Join(tmpDir, "mount")
+	stagePath := path.Join(tmpDir, "stage")
+
+	fakeMounter, fakeCloud, fakeMetadataService, fakeKubernetesClient := createMockObjects(mockCtrl)
+
+	mockNodeService(fakeMounter, fakeMetadataService)
+	mockControllerService(fakeCloud, mountPath)
+
+	driverOptions := &d.Options{
+		Mode:                              d.AllMode,
+		ModifyVolumeRequestHandlerTimeout: 60,
+		Endpoint:                          endpoint,
+	}
+
+	drv, err := d.NewDriver(fakeCloud, driverOptions, fakeMounter, fakeMetadataService, fakeKubernetesClient)
+	if err != nil {
+		t.Fatalf("Failed to create fake driver: %v", err.Error())
+	}
+	go func() {
+		if err := drv.Run(); err != nil {
+			panic(fmt.Sprintf("%v", err))
+		}
+	}()
+
+	config := csisanity.TestConfig{
+		TargetPath:                  mountPath,
+		StagingPath:                 stagePath,
+		Address:                     endpoint,
+		DialOptions:                 []grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())},
+		IDGen:                       csisanity.DefaultIDGenerator{},
+		TestVolumeSize:              10 * util.GiB,
+		TestVolumeAccessType:        "mount",
+		TestVolumeMutableParameters: map[string]string{"iops": "3014", "throughput": "153"},
+	}
+	csisanity.Test(t, config)
+}
+
+func createMockObjects(mockCtrl *gomock.Controller) (*mounter.MockMounter, *cloud.MockCloud, *metadata.MockMetadataService, *d.MockKubernetesClient) {
+	fakeMounter := mounter.NewMockMounter(mockCtrl)
+	fakeCloud := cloud.NewMockCloud(mockCtrl)
+	fakeMetadataService := metadata.NewMockMetadataService(mockCtrl)
+	fakeKubernetesClient := d.NewMockKubernetesClient(mockCtrl)
+
+	return fakeMounter, fakeCloud, fakeMetadataService, fakeKubernetesClient
+}
+
+func mockNodeService(m *mounter.MockMounter, md *metadata.MockMetadataService) {
```

Comment:
Is there a reason you prefer relying heavily on mocks within these two mega-mock functions in sanity_test.go (for node/controller service), instead of creating fake implementations similar to mount_unsupported https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/pkg/mounter/mount_unsupported.go? 

Perhaps this was was cleaner than the alternative? But the presence of so many `m.Expect().<Some mounter function>(gomock.Any()).DoAndReturn(...)` piqued my interest.

Reply:
No the only reason I had for doing so was to keep in party with previous CSI Sanity tests Eddie implemented. I would be more than happy to add a fake_sanity_node.go and fake_sanity_controller.go if folks think this is cleaner.

Reply:
+1. Fakes are easier to maintain and lot cleaner.

Reply:
Perfect will do in next rev

Reply:
> No the only reason I had for doing so was to keep in party with previous CSI Sanity tests Eddie implemented

Ah, IIRC poor eddie had to do that because he implemented those tests before his lovely mounter package refactor.

Reply:
Correct - mocks were used in the previous implementation due to our code structure at the time. Using stubs now is feasible, and as per the recommendation above would result in cleaner code.

Reply:
Correct - the previous implementation relied on mocks due to our code structure at the time. It should be feasible to use stubs instead as per the recommendation above which would be cleaner.